huaweicloud/resource_huaweicloud_rds_read_replica_instance.go

```diff
@@ -0,0 +1,644 @@
+	return nil
+}
+
+func getRdsInstanceByID(client *golangsdk.ServiceClient, instanceID string) (*instances.RdsInstanceResponse, error) {
+	listOpts := instances.ListRdsInstanceOpts{
+		Id: instanceID,
+	}
+	pages, err := instances.List(client, listOpts).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("An error occured while querying rds read replica instance %s: %s", instanceID, err)
+	}
+
+	resp, err := instances.ExtractRdsInstances(pages)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceList := resp.Instances
+	if len(instanceList) == 0 {
+		// return an empty rds instance
+		log.Printf("[WARN] can not find the specified rds read replica instance %s", instanceID)
+		instance := new(instances.RdsInstanceResponse)
+		return instance, nil
+	}
+
+	if len(instanceList) > 1 {
+		return nil, fmt.Errorf("retrieving more than one rds read replica instance by %s", instanceID)
+	}
+	if instanceList[0].Id != instanceID {
+		return nil, fmt.Errorf("the id of rds read replica instance was expected %s, but got %s",
+			instanceID, instanceList[0].Id)
+	}
+
+	return &instanceList[0], nil
+}
+
+func readAvailabilityZone(resp instances.RdsInstanceResponse) string {
+	node := resp.Nodes[0]
+	return node.AvailabilityZone
+}
+
+func readPrimaryInstanceID(resp instances.RdsInstanceResponse) (string, error) {
+	relatedInst := resp.RelatedInstance
+	for _, relate := range relatedInst {
+		if relate.Type == "replica_of" {
+			return relate.Id, nil
+		}
+	}
+	return "", fmt.Errorf("Error when get primary instance id for replica %s", resp.Id)
+}
+
+func resourceDiskEncryptionID(d *schema.ResourceData) string {
+	var encryptionID string
+	volumeRaw := d.Get("volume").([]interface{})
+
+	if len(volumeRaw) == 1 {
+		encryptionID = volumeRaw[0].(map[string]interface{})["disk_encryption_id"].(string)
+	}
+
+	return encryptionID
+}
+
+func resourceReplicaInstanceVolume(d *schema.ResourceData) *instances.Volume {
+	var volume *instances.Volume
+	volumeRaw := d.Get("volume").([]interface{})
+
+	if len(volumeRaw) == 1 {
+		volume = new(instances.Volume)
+		volume.Type = volumeRaw[0].(map[string]interface{})["type"].(string)
+		volume.Size = volumeRaw[0].(map[string]interface{})["size"].(int)
+		// the size is optional and invalid for replica, but it's required in sdk
+		// so just set 100 if not specified
+		if volume.Size == 0 {
+			volume.Size = 100
+		}
+	}
+	log.Printf("[DEBUG] volume: %+v", volume)
+	return volume
+}
+
+func rdsInstanceStateRefreshFunc(client *golangsdk.ServiceClient, instanceID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		instance, err := getRdsInstanceByID(client, instanceID)
+		if err != nil {
+			return nil, "FOUND ERROR", err
+		}
+		if instance.Id == "" {
+			return instance, "DELETED", nil
+		}
+
+		return instance, instance.Status, nil
+	}
+}
+
+func resourceReadReplicaInstanceImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```

Comment:
If we can get the `primary_instance_id` from read method, then there is no necessary to get it again.

Reply:
Yes, I have modified about this.